File: JavaScript/maxnumberof-K-sum-pairs.js
const maxOperations = function(nums, k) {
    
    let map = new Map()
    let count = 0;

    for(let ch of nums){

        let y = k - ch

        if(map.get(y)>0){
            count++
            map.set(y, map.get(y)-1)
        }
        else{
            map.set(ch, (map.get(ch) || 0)+1)
        }
    }
    return count
};

File: Python/MaxNumberofK-SumPairs.py
class Solution:
    def maxOperations(self, nums: list[int], k: int) -> int:

        nums.sort()

        i = 0
        j = len(nums) - 1
        count = 0
        while i < j:
            total = nums[i] + nums[j]

            if total == k:
                count += 1
                i += 1
                j -= 1
            elif total > k:
                j -= 1
            else:
                i += 1

        return count


solution = Solution()
print(solution.maxOperations([1, 2, 3, 4], 5))
